pkg/embeddedcluster/monitor.go

```diff
@@ -0,0 +1,110 @@
+package embeddedcluster
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/replicatedhq/embedded-cluster-operator/api/v1beta1"
+	"github.com/replicatedhq/kots/pkg/logger"
+	"github.com/replicatedhq/kots/pkg/store"
+	"k8s.io/client-go/kubernetes"
+)
+
+var stateMut = sync.Mutex{}
+
+// MaybeStartClusterUpgrade checks if the embedded cluster is in a state that requires an upgrade. If so,
+// it starts the upgrade process. We only start an upgrade if the following conditions are met:
+// - The app has an embedded cluster configuration.
+// - The app embedded cluster configuration differs from the current embedded cluster config.
+func MaybeStartClusterUpgrade(ctx context.Context, client kubernetes.Interface, store store.Store, conf *v1beta1.Config) error {
+	if conf == nil {
+		return nil
+	}
+
+	isEC, err := IsEmbeddedCluster(client)
+	if err != nil {
+		return fmt.Errorf("failed to check if embedded cluster is enabled: %w", err)
+	}
+	if !isEC {
+		return nil
+	}
+
+	spec := conf.Spec
+	if upgrade, err := RequiresUpgrade(ctx, spec); err != nil {
+		return fmt.Errorf("failed to check if upgrade is required: %w", err)
+	} else if !upgrade {
+		return nil
+	}
+	if err := startClusterUpgrade(ctx, spec); err != nil {
+		return fmt.Errorf("failed to start cluster upgrade: %w", err)
+	}
+
+	go watchClusterState(ctx, store)
+
+	return nil
+}
+
+// InitClusterState initializes the cluster state in the database. This should be called when the
+// server launches.
+func InitClusterState(ctx context.Context, client kubernetes.Interface, store store.Store) error {
+	isEC, err := IsEmbeddedCluster(client)
+	if err != nil {
+		return fmt.Errorf("failed to check if embedded cluster is enabled: %w", err)
+	}
+	if isEC {
+		_, err = updateClusterState(ctx, store, "")
+		return err
+	}
+	return nil
+}
+
+// watchClusterState checks the status of the installation object and updates the cluster state
+// after the cluster state has been 'installed' for 30 seconds, it will exit the loop.
+// this function is blocking and should be run in a goroutine.
+// if it is called multiple times, only one instance will run.
+func watchClusterState(ctx context.Context, store store.Store) {
+	stateMut.Lock()
```

Comment:
do we need to unlock this when the function finishes so it could run again, if needed?

Reply:
Whoops

Reply:
fixed